\chapter{Introduction}

Obtaining DNA sequence of an organism is a difficult problem
as modern sequencing technologies can only "read" small pieces
of the genome (called reads). 
These reads are joined together using programs called genome assemblers
in the process called genome assembly.
Nowadays we have a surprising variety of sequencing technologies,
each with different characteristics. For example,
Illumina platforms have high accuracy, but produce short reads. 
On the other hand, Pacific Biosciences offer long reads with high number
of errors. 

Also there are many different assemblers with different algorithmic
approaches and not all of them can cope with variety of possible input data.
Also some of the assemblers produce assemblies with errors, so
there is a need for validation of the assemblies.

In this work we will review most used algorithm for genome assembly,
genome assembly validation and probablistic frameworks for evaluation
genome assembly likelihood. We will propose new methods
for genome assembly and solve algorithmic problems associated with them.

\section{Structure}

The structure of this document is following. In the rest of this chapter we
describe necessary biological terms and define several notation used throughout
the text.

In the second chapter we will discuss the genome assembly
problem and algorithms for genome assembly.
We will also mention several formalizations of genome assembly problem.

The third chapter describes frameworks for evaluating genome assembly.
We will describe several algorithms which find errors in the assembly
using simple statistics. Also we will describe several models
for evaluation likelihood of an assembly, which can be used
to compare assemblies.

The fifth chapter contains our PhD project with the proposed directions for our
research. We want to focus on using probabilistic framework 
for genome assembly and on algorithm problems which arise during usage this framework.

\section{Biological Introduction}

%\todo{zmen nazov kapitoly}

%DNA,RNA,protein, amino-acid, residue, base, gene -- exon, conod, start codon,
%stop codon, cDNA, intron, splice site, donor, acceptor, intergenic region,
%homolog,

In this section we review several biological terms that will be needed. More
information about DNA, proteins and genes can be found in
\cite{BiologyForDummies,UnderstandingBioinformatics}.  Every cell of living
organisms contains one or several \firstUseOf{DNA} molecules. DNA is a double
stranded molecule, consisting of two long sequences of \firstUseOf{nucleotides}
(nucleotides are also referred to as bases or \firstUseOf{residues}). There are
four types of nucleotides in DNA: adenosine, cytosine, guanine and thymine
represented by letters $A,C,G$ and $T$ respectively. In RNA nucleotide $T$ is
replaced with uracil, denoted by $U$. The strands in DNA are complementary. The
nucleotides at the same position in two strands are connected by hydrogen bonds
and are complementary: $A$ is always connected with $T$ and $C$ is always
connected with $G$. Therefore we can represent DNA by a sequence over alphabet
$\{A,C,G,T\}$. The complementary strand can be easily computed.

DNA encodes \firstUseOf{proteins}. Proteins play an important role in cell
biology since  they regulates many processes in the cell and are catalysts to
many chemical reactions. Proteins are sequences of \firstUseOf{amino acid}
molecules. There are $20$ amino acids that can be encoded in DNA. Parts of DNA
that encode proteins are called \firstUseOf{genes} (gene is ``substring'' of DNA
that will be translated into one protein).

Every gene consists of two types of sequences: exons which encode amino-acids,
and introns are removed before translation (some genes do no have any introns).
At first, double stranded DNA is copied or transcribed into single stranded RNA
called mRNA. Then, the introns are spliced out of mRNA and exons are merged into
one sequence: spliced mRNA. This sequence is then translated into a sequence of
amino-acids (protein). Each amino acid is encoded by a triplet of nucleotides
called \firstUseOf{codon}. Amino acids have multiple representations: different
codons can encode the same amino acid. Since codons are always triplets, the
length of coding region in the  spliced mRNA is multiple of three. However, the
length of individual exons does not have to be multiple of three since introns
can be located inside codons.  The coding sequence of a gene begins with a start
codon (sequence $ATG$) and the last codon of a gene is called stop codon
(sequences $TAG,TAA$ or $TGA$).

An intron begins with 5' splice site (also called donor site),  and ends with 3'
splice site (or acceptor site).  Acceptor and donor sites are sequence motifs of
fixed length \cite{Pairagon2009,UnderstandingBioinformatics}.


\section{Notations}

In this section we summarize notation used in the following
chapters.

All sequences, members of sets, vectors, and rows and columns of matrices will
be indexed from $0$. We will use mostly right-open intervals: $I=[a,b)$ means
that $a\in I$, but $b\notin I$. 

The element at the $k$-th position (zero based) of string (or sequence) $s$ will
be written as $s[k]$. The substring $s[i]s[i+1]\dots s[j-1]$ is denoted
$s[i:j]$.  If $n$ is the length of the string $s$ then $s[:i]$ is equivalent to
$s[0:i]$ and $s[i:]$ is equivalent to $s[i:n]$.  We will use the terms sequence
and string interchangeably.

Let $M$ be a matrix. Then $M[i,j]$ is the element from the $i$-th row and $j$-th
column of $M$ (indices are zero based). Similarly as for strings, submatrix
$M[i:j,k:l]$ is a matrix consisting from the intersection of rows $i,i+1,\dots,
j-1$ with columns $k,k+1,\dots,l-1$. If $M$ is of size $n\times m$ then
$M[:i,j:]$ is equivalent to $M[0:i,j:m]$.  The term $M[i,:]$ is equivalent to
$M[i,0:m]$ which is the $i$-th row of $M$.
